File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/LogicalNot.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class LogicalNot extends DynamicCustomOp {

    public LogicalNot(){ }

    public LogicalNot(SameDiff sd, SDVariable in1, SDVariable in2){
        super(null, sd, new SDVariable[]{in1, in2});
    }

    @Override
    public String opName(){
        return "boolean_not";
    }

    @Override
    public String tensorflowName() {
        return "LogicalNot";
    }


    @Override
    public List<SDVariable> doDiff(List<SDVariable> f1) {
        return Arrays.asList( sameDiff.zerosLike(larg()), sameDiff.zerosLike(rarg()));
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes){
        Preconditions.checkState(dataTypes != null && dataTypes.size() == 1, "Expected exactly 1 input datatype, got %s", dataTypes);
        Preconditions.checkState(dataTypes.get(0) == DataType.BOOL, "Datatype for input must be BOOL: got %s", dataTypes.get(0));
        return Collections.singletonList(DataType.BOOL);
    }
}

File: libnd4j/include/ops/declarable/platform/armcompute/logical_not.cpp
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

#include <ops/declarable/OpRegistrator.h>
#include <ops/declarable/PlatformHelper.h>
#include <system/platform_boilerplate.h>

#include "armcomputeUtils.h"

namespace sd {
namespace ops {
namespace platforms {

//////////////////////////////////////////////////////////////////////////
PLATFORM_IMPL(boolean_not, ENGINE_CPU) {
  auto x = INPUT_VARIABLE(0);
  auto output = OUTPUT_VARIABLE(0);

  Arm_TensorShape xShape, outShape;
  xShape.set_num_dimensions(x->rankOf());
  outShape.set_num_dimensions(output->rankOf());

  for (int i = 0; i < x->rankOf(); i++) {
    xShape[x->rankOf() - 1 - i] = x->sizeAt(i);
  }
  for (int i = 0; i < output->rankOf(); i++) {
    outShape[output->rankOf() - 1 - i] = output->sizeAt(i);
  }

  Arm_TensorInfo xInfo(xShape, 1, arm_compute::DataType::U8);
  Arm_TensorInfo outInfo(outShape, 1, arm_compute::DataType::U8);

  Arm_Tensor xTensor, outTensor;
  xTensor.allocator()->init(xInfo);
  outTensor.allocator()->init(outInfo);

  arm_compute::NELogicalNot logicalNot;
  logicalNot.configure(&xTensor, &outTensor);

  xTensor.allocator()->allocate();
  outTensor.allocator()->allocate();

  auto xPtr = reinterpret_cast<uint8_t*>(xTensor.buffer());
  for (sd::LongType i = 0; i < x->lengthOf(); i++) {
    xPtr[i] = x->e<bool>(i) ? 1 : 0;
  }

  logicalNot.run();

  auto outPtr = reinterpret_cast<uint8_t*>(outTensor.buffer());
  for (sd::LongType i = 0; i < output->lengthOf(); i++) {
    output->p(i, outPtr[i] != 0);
  }

  return sd::Status::OK;
}

//////////////////////////////////////////////////////////////////////////
PLATFORM_CHECK(boolean_not, ENGINE_CPU) {
  auto x = INPUT_VARIABLE(0);

  Requirements req("ARMCOMPUTE BOOLEAN_NOT OP");
  req.expectEq(makeInfoVariable(x->dataType(), TYPE_MSG_INPUT), DataType::BOOL) &&
      req.expectLessEq(makeInfoVariable(x->rankOf(), RANK_MSG_INPUT), arm_compute::MAX_DIMS) &&
      req.expectGreater(makeInfoVariable(x->rankOf(), RANK_MSG_INPUT), 0) &&
      req.expectEq(makeInfoVariable(x->ordering(), ORDERING_MSG_INPUT), 'c');
  req.logTheSuccess();
  return req;
}

}  // namespace platforms
}  // namespace ops
}  // namespace sd
